File: solutions/dart/list-ops/1/lib/list_ops.dart
// Soulfully offered to Shree DR.MDD 🌼

extension ListOps on List {
  U foldl<T, U>(U seed, U Function(U acc, T val) combine) {
    var aggregate = seed;
    forEach((entry) => aggregate = combine(aggregate, entry));
    return aggregate;
  }

  int count() => foldl(0, (total, _) => total + 1);

  void append<T>(List<T> seq) => seq.forEach(add);

  List<T> concat<T>() => foldl([], (merged, List<T> unit) {
        merged.append(unit);
        return merged;
      });

  List<T> filter<T>(bool Function(T item) test) =>
      foldl([], (picked, T item) {
        if (test(item)) picked.add(item);
        return picked;
      });

  List<T> myMap<T>(T Function(T item) mapper) => foldl([], (altered, T item) {
        altered.add(mapper(item));
        return altered;
      });

  List<T> reverse<T>() => foldl([], (inverted, T item) {
        inverted.insert(0, item);
        return inverted;
      });

  U foldr<T, U>(U seed, U Function(T item, U acc) merge) =>
      reverse().foldl(seed, (U acc, T item) => merge(item, acc));
}

File: solutions/erlang/list-ops/1/src/list_ops.erl
-module(list_ops).
-compile({no_auto_import, [length/1]}).
-export([append/2, concat/1, filter/2, length/1, map/2, foldl/3, foldr/3,
	 reverse/1]).
append([], L2) -> L2;
append([E|L1], L2) -> [E|append(L1, L2)].
concat([]) -> [];
concat([E|L]) -> append(E, concat(L)).
filter(_, []) -> [];
filter(F, [E|L]) -> case F(E) of true -> [E|filter(F, L)]; false -> filter(F, L) end.
length([]) -> 0;
length([_|L]) -> 1+length(L).
map(_, []) -> [];
map(F, [E|L]) -> [F(E)|map(F, L)].
foldl(_, Acc, []) -> Acc;
foldl(F, Acc, [E|L]) -> foldl(F, F(E, Acc), L).
foldr(F, S, L) -> foldl(F, S, reverse(L)).
reverse(L) -> reverse(L, []).
reverse([], Acc) -> Acc;
reverse([E|L], Acc) -> reverse(L, [E|Acc]).